src/Cli/dotnet/Commands/Tool/Run/ToolRunFromSourceCommand.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.CommandLine;
+using Microsoft.DotNet.Cli.CommandFactory;
+using Microsoft.DotNet.Cli.CommandFactory.CommandResolution;
+using Microsoft.DotNet.Cli.Commands.Tool.Install;
+using Microsoft.DotNet.Cli.ToolPackage;
+using NuGet.Versioning;
+
+namespace Microsoft.DotNet.Cli.Commands.Tool.Run
+{
+    internal class ToolRunFromSourceCommand(ParseResult result) : CommandBase(result)
+    {
+        private readonly string _toolCommandName = result.GetValue(ToolRunCommandParser.CommandNameArgument);
+        private readonly IEnumerable<string> _forwardArguments = result.GetValue(ToolRunCommandParser.CommandArgument);
+        private readonly bool _allowRollForward = result.GetValue(ToolRunCommandParser.RollForwardOption);
+        private readonly string _configFile = result.GetValue(ToolRunCommandParser.FromSourceConfigFile);
+        private readonly string[] _sources = result.GetValue(ToolRunCommandParser.FromSourceSourceOption) ?? [];
+        private readonly string[] _addSource = result.GetValue(ToolRunCommandParser.FromSourceAddSourceOption) ?? [];
+        private readonly bool _ignoreFailedSources = result.GetValue(ToolCommandRestorePassThroughOptions.IgnoreFailedSourcesOption);
+        private readonly bool _interactive = result.GetValue(ToolRunCommandParser.FromSourceInteractiveOption);
+        private readonly VerbosityOptions _verbosity = result.GetValue(ToolRunCommandParser.FromSourceVerbosityOption);
+        private readonly bool _yes = result.GetValue(ToolRunCommandParser.FromSourceYesOption);
+
+        public override int Execute()
+        {
+            if (!UserAgreedToRunFromSource())
+            {
+                return 1;
```

Comment:
Should we instead return the exception that would get returned if the package is not found?